openlibrary/catalog/utils/__init__.py

```diff
@@ -319,3 +324,111 @@ def expand_record(rec: dict) -> dict[str, str | list[str]]:
         if f in rec:
             expanded_rec[f] = rec[f]
     return expanded_rec
+
+
+def get_publication_year(publish_date: str | int | None) -> int | None:
+    """
+    Return the publication year from a book in YYYY format by looking for four
+    consecutive digits not followed by another digit. If no match, return None.
+
+    >>> get_publication_year('1999-01')
+    1999
+    >>> get_publication_year('January 1, 1999')
+    1999
+    """
+    if publish_date is None:
+        return None
+
+    pattern = compile(r"\b\d{4}(?!\d)\b")
+    match = pattern.search(str(publish_date))
+
+    return int(match.group(0)) if match else None
+
+
+def published_in_future_year(publish_year: int) -> bool:
+    """
+    Return True if a book is published in a future year as compared to the
+    current year.
+
+    Some import sources have publication dates in a future year, and the
+    likelihood is high that this is bad data. So we don't want to import these.
+    """
+    return publish_year > datetime.datetime.now().year
```

Comment:
@seabelis I'm wondering if we might want to add a bit of wiggle room here. I find a lot of valid books are imported a little before their release date. Eg https://openlibrary.org/search?q=first_publish_year%3A2024&mode=everything

Currently it will block imports with publication year greater than eg 2023. Do you think we should allow for a +1 margin here?

Reply:
@cdrini To my way of looking at it we should not import something that isn’t yet published. Editions that were once available but are no longer extant are a more subtle call, but I’d avoid auto-importing them. We already have a huge pile of crap to clean up from bogus POD that were put out by “independently published” or the many others of that ilk.